### PR TITLE
Refactor/population panel constructor

### PR DESCRIPTION
**Pull request by oscar139**
Initial attempt to address issue #1442.

### CODE REVIEW

File: libOPHD/Population/PopulationPool.h
Comment (oscar139) [inline libOPHD/Population/PopulationPool.h:28]: scientistsAsWorkers, populationEmployed, and size have a similar behavior as the other functions. Does it make sense to change them too. I made the changes locally and it did not appear to create any issues.

Comment (DanRStevens) [reply libOPHD/Population/PopulationPool.h:28]: It's often a good idea to marking functions as `const` as a sort of default, and only remove it if the function really needs to mutate state.

A quick note on semantics: I've seen cases where you could technically mark a function as `const`, and it would compile, but that was only because there was a level of indirection to the data being modified. The internal level of indirection is an implementation detail, and the nature of the function in that case really was to mutate state. In such cases, I think it makes the most sense to not declare a method as `const`. In particular, if you were to refactor the internal implementation, you might be required to remove the `const`, which could potentially break calling code if it assumed a method was callable on a `const` object.

Comment (DanRStevens) [reply libOPHD/Population/PopulationPool.h:28]: It doesn't really make sense to declare the return values as `const`. They are being copied anyway.

Besides, the caller should decide if the result they're storing should be `const` or not.

File: OPHD/UI/PopulationPanel.h
Comment (DanRStevens) [inline OPHD/UI/PopulationPanel.h:60]: I generally recommend against `init` methods, as what they contain should really be in a constructor.

These did make sense in the past, when you wanted to share code between constructors, which I see is the direction you were going here. Since C++11 this approach is no longer necessary, since "Delegating Constructors" were introduced to the language. When you want to re-use constructors, you can call one from the other using the member initialization list. In that case, only that one single "member" can be initialized.

Some details are found about half way down this page:
https://en.cppreference.com/w/cpp/language/constructor

One approach might be for the default constructor to call a constructor taking parameters, and passing default values. However, in that case, it might make sense to set default values for the parameters, making them optional, and allowing that constructor to also function as the default constructor.

In this particular case, that's possible if you're still using pointer syntax, but not reference syntax. You can't quite switch to reference syntax until after the setters are removed, since a reference is basically a `const` pointer (to data that may or may not be `const`).

**Review by DanRStevens — COMMENTED** [review]
Looks like there are merge conflicts that need to be resolved. Probably from when PR #1434 was merged.

File: OPHD/UI/PopulationPanel.h
Comment (ldicker83) [reply OPHD/UI/PopulationPanel.h]: Generally agreed here -- Part of the style that I've been using lately is "Only 5 Lines of Code" per function (including c'tor) and I tend to extract some of that code into stand alone functions that are wrapped in an anonymous namespace.

Comment (DanRStevens) [reply OPHD/UI/PopulationPanel.h]: My general rule of thumb is to try and keep functions within one page of code, so you never need to scroll to see the full context. There's a lot less you need to remember when you can see it all at once.

File: OPHD/UI/PopulationPanel.cpp
Comment (DanRStevens) [inline OPHD/UI/PopulationPanel.cpp]: These two lines should probably have been part of the previous commit. It might also have been helpful to use the constructor init-list, rather than the constructor body. That would be important for the switch to reference syntax, since the constructor body is too late to initialize references (or `const` values).

Note that a pointer to const data is not itself const. The pointer can still be modified, even if the data is const. A reference is like a const pointer, in that the pointer itself can not be modified after initialization.

File: OPHD/States/MapViewStateUi.cpp
Comment (DanRStevens) [inline OPHD/States/MapViewStateUi.cpp:79]: This could potentially have been grouped into the previous commit. Once initialization is done using constructor parameters, there is no longer a need to call the setter methods. Doing so initializes the values twice.

File: OPHD/UI/PopulationPanel.cpp
Comment (DanRStevens) [inline OPHD/UI/PopulationPanel.cpp:70]: Ahh, I see internally pointer syntax is still being used. For a moment I was wondering why this didn't result in a compiler error. If reference syntax had been used internally, then this initialization would have been too late. The constructor init list is needed to initialize references. Once the constructor body is entered, all component objects are already initialized.

Or was that why pointer syntax was still being used internally?

**Review by DanRStevens — APPROVED** [review]
Very nicely presented. Very easy to read through and understand.

For the second commit, there were two changes grouped together, which could possibly have been separate commits.
- Changing from calling setter methods to using new constructor parameters
- Marking the fields and parameters as `const`

Though that's a fairly minor observation.

I've pushed the branch to the main repo so the CI build will run. I'll merge this once the builds finish, assuming no problems.

----

Edit: Huh, I see CI build failures, though the logs make it look like it actually passed. I may need to investigate a bit further to see what happened.

**Comment by DanRStevens** [comment]
Looks like constructor init list items don't match the order of the fields in the class. We have a warning flag enabled that checks for that.

Member fields are always initialized in their declaration order, not the order they appear in the init list. Hence putting them out of order in the init list can be misleading. One of the reasons for that is a class can have many constructors, which means it can have many init list orderings. The class definition only appears once though, so the order specified there is what is used.

Initialization order is important in the case of exceptions, since the exception handler must destruct all objects in the reverse order they were constructed (and only for the objects that completed construction). If a constructor fails with an exception part way through, some set of member fields may need to be destructed, depending on how far into construction the class got.

I'm not entirely sure why all constructors need to initialize objects in the same order, though if all constructors initialize objects in the same order, than they can all share the same exception handling code to destruct objects in the same order.

**Review by DanRStevens — APPROVED** [review]

**Pull request merged**
